### ISSUE TITLE
Misc. bug: The cache_prompt parameter not working properly

### ISSUE THREAD

**Issue by feikiss** [ggerganov/llama.cpp#10993]
### Name and Version

version: 4149 (1bb30bf2)
built with cc (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0 for x86_64-linux-gnu

### Operating systems

Linux

### Which llama.cpp modules do you know to be affected?

llama-server

### Problem description & steps to reproduce

The param "cache_prompt" seems not working as expected. 
I'm using the llama.cpp in pure CPU env. Command to start the server: `/llama-server -m  /models/qwen2.5-1.5b-q8/qwen2.5-1.5b-instruct-q8_0.gguf -c 1024 --host 0.0.0.0 --port 8000 -dkvc --metrics --file  /models/promts/tool.txt --keep -1`
The URI I'm using: `/v1/completions` , body is as below:
```
{
"prompt":"LONG_PROMPT + short_question1,
"cache_prompt": true,
xxxx,// other params
}
```
If I invoke the API with the prompt format "LONG_PROMPT" + short_question again and again, it can always work well, and reply quickly, but when I asked one another prompt, just like "short_question2", then back to call with "LONG_PROMPT" + short_question1， it will coast much a long time. It seems the prompt cache for the "LONG_PROMPT" is lost if I change the prompt without the "LONG_PROMPT".

**Comment by ggerganov** [comment]
This is the expected behavior - the caching only works with the last prompt. You can use slot save/load API for more advanced caching.

**Comment by feikiss** [comment]
> This is the expected behavior - the caching only works with the last prompt. You can use slot save/load API for more advanced caching.

Thanks @ggerganov .  What's the perpose or scenario  that just caching the last prompt?  As my test,  I cannot control which slot to use while calling the API (/v1/completions) , so in the last all slots will be relpaced with the last prompt.

**Comment by ggerganov** [comment]
> Thanks @ggerganov . What's the perpose or scenario that just caching the last prompt?

The typical use case is a chat session - the client appends questions to the existing context and it keeps growing. Thanks to prompt caching, the server will compute only the tokens of the new question and reuse all previous tokens.

**Comment by feikiss** [comment]
> > Thanks @ggerganov . What's the perpose or scenario that just caching the last prompt?
> 
> The typical use case is a chat session - the client appends questions to the existing context and it keeps growing. Thanks to prompt caching, the server will compute only the tokens of the new question and reuse all previous tokens.

Yes, it is meaningful if just hold one chat session, but how to handle two or more different chat sessions at the same time? Can I control each chat session use specific  slot seperately?

**Comment by ggerganov** [comment]
The design is that you can distribute the different users to different slots and each slot will cache it's own chat session. When a user ends a chat session, then you can put another user on that slot. For example, if you are going to have a maximum of 4 parallel users, then start the server with `-np 4` and assign each user to a different slot.

**Comment by feikiss** [comment]
Thanks @ggerganov ! My queston is "how to assign each user to a different slot ?"  I have read the[ server docs](https://github.com/ggerganov/llama.cpp/tree/master/examples/server)  again and again, but not found the answer. Thanks.
